### PR TITLE
fix dead lock in concurrent search/release segment

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>
issue: #26428

### CODE REVIEW

**Comment by yah01** [comment]
The `GetAndPinBy` and `GetAndPin` methods need to be changed to acquiring mutex lock, or the deadlock will still exist:
1. 2 goroutines are calling `GetAndPin`
2. another goroutine is unpinning segments
3. the order of segments being pinning in the 2 goroutines is diff

**Comment by weiliu1031** [comment]
> The `GetAndPinBy` and `GetAndPin` methods need to be changed to acquiring mutex lock, or the deadlock will still exist:
> 
> 1. 2 goroutines are calling `GetAndPin`
> 2. another goroutine is unpinning segments
> 3. the order of segments being pinning in the 2 goroutines is diff
pin segment used read lock, won't block each other, so won't get deadlock here

File: internal/querynodev2/segments/manager.go [hunk 1]
```diff
@@ -388,64 +385,116 @@ func (mgr *segmentManager) Empty() bool {
...
+		}
+
+	case SegmentTypeSealed:
+		s, ok := mgr.sealedSegments[segmentID]
+		if ok {
+			delete(mgr.sealedSegments, segmentID)
+			return s.(*LocalSegment)
+		}
+	default:
+		return nil
```
Comment (yah01) [inline internal/querynodev2/segments/manager.go:445]: we could just remove the default branch, maybe next PR

**Review by yah01 — APPROVED** [review]
/lgtm

File: internal/querynodev2/segments/manager.go [hunk 2]
```diff
@@ -388,64 +385,116 @@ func (mgr *segmentManager) Empty() bool {
...
+	for id, segment := range mgr.growingSegments {
+		delete(mgr.growingSegments, id)
+		remove(segment.(*LocalSegment))
 	}
 
-	for id := range mgr.sealedSegments {
-		remove(id, mgr.sealedSegments)
+	for id, segment := range mgr.sealedSegments {
+		delete(mgr.sealedSegments, id)
+		remove(segment.(*LocalSegment))
```
Comment (congqixia) [inline internal/querynodev2/segments/manager.go:497]: if `Clear` & `Unpin` happens concurrently, this will always cause deadlock here

Comment (weiliu1031) [reply internal/querynodev2/segments/manager.go:497]: Unpin won't acquire  manager's lock here

**Pull request merged**